chain/client/src/stateless_validation/state_witness_producer.rs

```diff
@@ -163,88 +163,89 @@ impl Client {
 
     /// Collect state transition data necessary to produce state witness for
     /// `chunk_header`.
+    /// Returns main state transition and implicit transitions, in the order
+    /// they should be applied, and the hash of receipts to apply.
     fn collect_state_transition_data(
         &mut self,
         chunk_header: &ShardChunkHeader,
         prev_chunk_header: &ShardChunkHeader,
     ) -> Result<StateTransitionData, Error> {
-        let store = self.chain.chain_store().store();
-        let shard_id = chunk_header.shard_id();
-        let epoch_id =
-            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
-        let shard_uid = self.epoch_manager.shard_id_to_uid(shard_id, &epoch_id)?;
         let prev_chunk_height_included = prev_chunk_header.height_included();
-        let mut prev_blocks = self.chain.get_blocks_until_height(
-            *chunk_header.prev_block_hash(),
-            prev_chunk_height_included,
-            true,
-        )?;
-        prev_blocks.reverse();
-        let (main_block, implicit_blocks) = prev_blocks.split_first().unwrap();
-        let (base_state, receipts_hash, contract_accesses) = if prev_chunk_header.is_genesis() {
-            (Default::default(), hash(&borsh::to_vec::<[Receipt]>(&[]).unwrap()), vec![])
-        } else {
-            let state_transition = store
-                .get_ser(
-                    near_store::DBCol::StateTransitionData,
-                    &near_primitives::utils::get_block_shard_id(main_block, shard_id),
-                )?
-                .ok_or_else(|| {
-                    let message = format!(
-                        "Missing main transition state proof for block {main_block} and shard {shard_id}"
-                    );
-                    if !cfg!(feature = "shadow_chunk_validation") {
-                        log_assert_fail!("{message}");
-                    }
-                    Error::Other(message)
-                })?;
-            match state_transition {
-                StoredChunkStateTransitionData::V1(StoredChunkStateTransitionDataV1 {
-                    base_state,
-                    receipts_hash,
-                    contract_accesses,
-                }) => (base_state, receipts_hash, contract_accesses),
-            }
-        };
-        let main_transition = ChunkStateTransition {
-            block_hash: *main_block,
-            base_state,
-            post_state_root: *self.chain.get_chunk_extra(main_block, &shard_uid)?.state_root(),
-        };
 
+        // Iterate over blocks in chain from `chunk_header.prev_block_hash()`
+        // (inclusive) until the block with height `prev_chunk_height_included`
+        // (exclusive).
+        // Every block corresponds to one implicit state transition between
+        // `prev_chunk_header` and `chunk_header`.
+        // There may be one additional implicit transition for a block, if
+        // resharding happens after its processing.
```

Comment:
That is a really nice observation, I like this way of thinking about resharding!